### PR TITLE
Fixed some memory leaking

### PR DESCRIPTION
**Pull request by skyline75489**
- [ ] This change has been discussed with the author.
- [X] It implements / fixes issue - general memory issues.

---

**Description:**

### CODE REVIEW

File: iina/FileGroup.swift [hunk 1]
```diff
@@ -17,7 +17,6 @@ class FileInfo: Hashable {
   var filename: String
   var ext: String
   var nameInSeries: String?
-  var characters: [Character]
```
Comment (lhc70000) [inline iina/FileGroup.swift]: Are you sure there's no performance issue caused by accessing `characters` and constructing `Index` in each iteration?

File: iina/HistoryWindowController.swift [hunk 2]
```diff
@@ -77,6 +77,10 @@ class HistoryWindowController: NSWindowController, NSOutlineViewDelegate, NSOutl
     outlineView.expandItem(nil, expandChildren: true)
   }
 
+  deinit {
+    NotificationCenter.default.removeObserver(self)
+  }
```
Comment (lhc70000) [inline iina/HistoryWindowController.swift]: Please wrap it with `ObjcUtils.silenced`.

It seems that Cocoa tends to release (leak) the whole memory block after quit and let system GC to handle it, so `deinit` is not being called at a fixed time. In early versions IINA crashes when trying to remove nonexistent observers. I'm not sure it's still needed but it's better to add them for safety reasons.

File: iina/FileGroup.swift (hunk 1 above)
Comment (skyline75489) [inline iina/FileGroup.swift]: I don't see any obvious performance issue at this point. The amount of data isn't large enough, after all.

File: iina/MPVController.swift [hunk 3]
```diff
@@ -342,8 +342,13 @@ class MPVController: NSObject {
     strArgs.insert(command.rawValue, at: 0)
     strArgs.append(nil)
     var cargs = strArgs.map { $0.flatMap { UnsafePointer<Int8>(strdup($0)) } }
+    defer {
```
Comment (saagarjha) [inline iina/MPVController.swift]: Why is a `defer` necessary?

File: iina/QuickSettingViewController.swift [hunk 4]
```diff
@@ -147,8 +147,10 @@ class QuickSettingViewController: NSViewController, NSTableViewDataSource, NSTab
   }
 
   deinit {
-    observers.forEach {
-      NotificationCenter.default.removeObserver($0)
+    ObjcUtils.silenced {
```
Comment (saagarjha) [inline iina/QuickSettingViewController.swift]:
```swift
guard #available(OSX 10.11, *) else {
  ObjcUtils.silenced { [unowned self] in
    self.observers.forEach(NotificationCenter.default.removeObserver)
  }
  return
}
```

File: iina/PlaylistViewController.swift [hunk 5]
```diff
@@ -102,7 +102,9 @@ class PlaylistViewController: NSViewController, NSTableViewDataSource, NSTableVi
   }
 
   deinit {
-    NotificationCenter.default.removeObserver(self.playlistChangeObserver!)
+    ObjcUtils.silenced {
```
Comment (saagarjha) [inline iina/PlaylistViewController.swift]:
```swift
guard #available(OSX 10.11, *) else {
  ObjcUtils.silenced { [unowned self] in
    self.playlistChangeObserver.flatMap { NotificationCenter.default.removeObserver($0) }
  }
  return
}
```

File: iina/AutoFileMatcher.swift [hunk 6]
```diff
@@ -45,9 +46,11 @@ class AutoFileMatcher {
 
     // group by extension
     for file in files {
-      let fileInfo = FileInfo(file)
-      if let mediaType = Utility.mediaType(forExtension: fileInfo.ext) {
-        filesGroupedByMediaType[mediaType]!.append(fileInfo)
+      autoreleasepool{
```
Comment (saagarjha) [inline iina/AutoFileMatcher.swift]: Why is this necessary?

File: iina/FilterWindowController.swift [hunk 7]
```diff
@@ -47,7 +47,9 @@ class FilterWindowController: NSWindowController {
   }
 
   deinit {
-    NotificationCenter.default.removeObserver(self)
+    ObjcUtils.silenced {
```
Comment (saagarjha) [inline iina/FilterWindowController.swift]:
```swift
guard #available(OSX 10.11, *) else {
  ObjcUtils.silenced { [unowned self] in
    NotificationCenter.default.removeObserver(self)
  }
  return
}
```

File: iina/HistoryWindowController.swift [hunk 8]
```diff
@@ -77,6 +77,12 @@ class HistoryWindowController: NSWindowController, NSOutlineViewDelegate, NSOutl
     outlineView.expandItem(nil, expandChildren: true)
   }
 
+  deinit {
+    ObjcUtils.silenced {
```
Comment (saagarjha) [inline iina/HistoryWindowController.swift]:
```swift
guard #available(OSX 10.11, *) else {
  ObjcUtils.silenced { [unowned self] in
    NotificationCenter.default.removeObserver(self)
  }
  return
}
```

File: iina/KeyRecordViewController.swift [hunk 9]
```diff
@@ -69,6 +69,12 @@ class KeyRecordViewController: NSViewController, KeyRecordViewDelegate, NSRuleEd
     }
   }
 
+  deinit {
+    ObjcUtils.silenced {
```
Comment (saagarjha) [inline iina/KeyRecordViewController.swift]:
```swift
guard #available(OSX 10.11, *) else {
  ObjcUtils.silenced { [unowned self] in
    NotificationCenter.default.removeObserver(self)
  }
  return
}
```

File: iina/FileGroup.swift [hunk 10]
```diff
@@ -144,8 +143,7 @@ class FileGroup {
 
   func flatten() -> [String: [FileInfo]] {
     var result: [String: [FileInfo]] = [:]
-    var search: ((FileGroup) -> Void)!
-    search = { group in
+    func search(_ group: FileGroup) {
```
Comment (skyline75489) [inline iina/FileGroup.swift]: There seems to be a bug inside Swift itself which causes memory leak when there's recursive call of closure. Using inner function prevents this kind of leak.

Comment (saagarjha) [inline iina/FileGroup.swift]: > There seems to be a bug inside Swift itself which causes memory leak when there's recursive call of closure.

Is there? Give me a few more details and I'll file it on the [Swift JIRA](http://bugs.swift.org/).

File: iina/MPVController.swift (hunk 3 above)
Comment (skyline75489) [inline iina/MPVController.swift]: AFAIK, `defer` is a good practice for releasing unsafe resources. Because you can use it right after the alloctaion code, it's more noticeable & error-resistant.

File: iina/AutoFileMatcher.swift (hunk 6 above)
Comment (skyline75489) [inline iina/AutoFileMatcher.swift]: Putting `autoreleasepool` inside loop is also a common practice to avoid memory spike, if that's what you asking.

File: iina/MPVController.swift (hunk 3 above)
Comment (saagarjha) [inline iina/MPVController.swift]: Yes, I know that. The resources are released two lines down, though.

File: iina/FileGroup.swift (hunk 10 above)
Comment (skyline75489) [inline iina/FileGroup.swift]: Have a look at instruments & you'll find out there's leak around here.

File: iina/FileGroup.swift (hunk 1 above)
Comment (skyline75489) [inline iina/FileGroup.swift]: This change is mainly because `self.characters = [Character](self.filename.characters)` is causing too much memory copying (and somehow shows leaking in Instruments). If you guys think performance has higher priority, I'm OK with reverting this.

File: iina/AutoFileMatcher.swift (hunk 6 above)
Comment (saagarjha) [inline iina/AutoFileMatcher.swift]: `autoreleasepool` is useful when you're working with `autorelease`d types. You're working with Swift `struct`s here.

File: iina/MPVController.swift (hunk 3 above)
Comment (skyline75489) [inline iina/MPVController.swift]: I'd like a consistent coding style with this kind of allocation. So I changed this on and the one in md5.

File: iina/AutoFileMatcher.swift (hunk 6 above)
Comment (skyline75489) [inline iina/AutoFileMatcher.swift]: Which `struct` are you talking about? `FileInfo` is a class here.

File: iina/FileGroup.swift (hunk 1 above)
Comment (saagarjha) [inline iina/FileGroup.swift]: Arrays are COW. Since no modifications are taking place no copying should be done.

Comment (skyline75489) [inline iina/FileGroup.swift]: That's my first thought but Instruments suggest otherwise. You know what, I'm beginning to doubt whether Instruments is pointing me to the right direction.

File: iina/FileGroup.swift (hunk 10 above)
Comment (skyline75489) [inline iina/FileGroup.swift]: At first I thought maybe `[weak search]`? But Swift complains about "'weak' may only be applied to class and class-bound protocol types, not '(FileGroup) -> Void'".

File: iina/AutoFileMatcher.swift (hunk 6 above)
Comment (saagarjha) [inline iina/AutoFileMatcher.swift]: Oops, I had the wrong `FileInfo`. However, my point still stands–`FileInfo` should not have to be `autorelease`d.

Comment (skyline75489) [inline iina/AutoFileMatcher.swift]: If the performance thing about character stands, it means there will be a lot of `FileInfo`. So...why not `autorelease` it?

Comment (saagarjha) [inline iina/AutoFileMatcher.swift]: IIRC Chris Lattner had said somewhere that the compiler will detect this and return an object with a +0 retain count in this case.

File: iina/MPVController.swift [hunk 11]
```diff
@@ -342,8 +342,13 @@ class MPVController: NSObject {
     strArgs.insert(command.rawValue, at: 0)
     strArgs.append(nil)
     var cargs = strArgs.map { $0.flatMap { UnsafePointer<Int8>(strdup($0)) } }
```
Comment (skyline75489) [inline iina/MPVController.swift]: This actually causes bad memory access for some reason.

Comment (saagarjha) [inline iina/MPVController.swift]: Oh wait, never mind. Leave it as is.

File: iina/AutoFileMatcher.swift (hunk 6 above)
Comment (skyline75489) [inline iina/AutoFileMatcher.swift]: Huh, in that case this isn't necessary. Is this thing applied to all loops? Forgive me cuz I'm mostly a objc guy.

**Comment by lhc70000** [comment]
All issues in this PR seems to be already resolved in current codebase. Closing.
